### PR TITLE
avoid n+1 in search

### PR DESCRIPTION
**Pull request by qnkhuat**
We have a n+1 query in our search api which we could avoid. this PR fixes that

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/28185)

### CODE REVIEW

File: test/metabase/api/search_test.clj [hunk 1]
```diff
@@ -609,3 +611,34 @@
...
+                           :metabase.models.segment/Segment      _             {:table_id table-id
+                                                                                :name     "segment count test 2"}
+                           :metabase.models.segment/Segment      _             {:table_id table-id
+                                                                                :name     "segment count test 3"}]
+    (toucan2.execute/with-call-count [call-count]
+      (#'api.search/search (#'api.search/search-context "count test" nil nil nil 100 0))
+      ;; the call count number here are expected to change if we change the search api
+      ;; we have this test here just to keep tracks this number to remind us to put effort
+      ;; into keep this number as low as we can
+      (is (= 7 (call-count))))))
```
Comment (qnkhuat) [inline test/metabase/api/search_test.clj:644]: it was 17 calls

File: src/metabase/api/search.clj [hunk 2]
```diff
@@ -405,21 +406,23 @@
...
-                           (map #(update % :bookmark api/bit->boolean))
-                           (map #(update % :archived api/bit->boolean))
-                           (map (partial scoring/score-and-result (:search-string search-ctx)))
-                           (filter #(pos? (:score %))))
-        total-results     (scoring/top-results reducible-results search-config/max-filtered-results xf)]
+  (let [search-query       (full-search-query search-ctx)
+        _                  (log/tracef "Searching with query:\n%s\n%s"
+                                       (u/pprint-to-str search-query)
+                                       (mdb.query/format-sql (first (mdb.query/compile search-query))))
+        to-toucan-instance (fn [o] (t2.instance/instance (search-config/model-to-db-model (:model o)) o))
```
Comment (qnkhuat) [inline src/metabase/api/search.clj:413]: added this fn to turn object from reducible-query to a fully qualified toucan2/instance so `mi/can-read?` could dispatch correctly.

Comment (qnkhuat) [reply src/metabase/api/search.clj:413]: @camsaul not sure if this idiomatic with toucan2? 

AFAIK, reducible-query can only have `modelable` for a single model?

File: src/metabase/api/search.clj [hunk 3]
```diff
@@ -151,7 +152,7 @@
 (s/defn ^:private from-clause-for-model :- [(s/one [(s/one s/Keyword "table name") (s/one s/Keyword "alias")]
                                                    "from clause")]
   [model :- SearchableModel]
-  (let [db-model (get search-config/model-to-db-model model)]
+  (let [db-model (get search-config/model-to-db-model-t1 model)]
```
Comment (calherries) [inline src/metabase/api/search.clj:155]: You should be able to use `search-config/model-to-db-model` here. `t2/table-name` will call `name` on a keyword.

File: src/metabase/search/config.clj [hunk 4]
```diff
@@ -37,8 +37,8 @@
   "Show this many words of context before/after matches in long search results"
   2)
 
-(def model-to-db-model
-  "Mapping from string model to the Toucan model backing it."
+(def model-to-db-model-t1
```
Comment (calherries) [inline src/metabase/search/config.clj:40]: You should be able to get rid of this, and only use model-to-db-model

File: src/metabase/api/search.clj (hunk 3 above)
Comment (qnkhuat) [reply src/metabase/api/search.clj:155]: per Cam's [message](https://metaboat.slack.com/archives/CKZEMT1MJ/p1675879200978579), looks like the preferable way is to use keyword with toucan2?

Comment (calherries) [reply src/metabase/api/search.clj:155]: I'm unsure what part of Cam's message you're talking about, but yes I'm advocating for using keywords and toucan2

Comment (calherries) [reply src/metabase/api/search.clj:155]: As in, you shouldn't need `model-to-db-model-t1` anymore. Just use the toucan2 model keywords

**Comment by calherries** [comment]
Looks good to me 👍 Not sure about [this](https://github.com/metabase/metabase/pull/28185#discussion_r1101604202) though

**Review by calherries — APPROVED** [review]

File: src/metabase/api/search.clj [hunk 5]
```diff
@@ -364,16 +365,16 @@
   true)
 
 (defmethod check-permissions-for-model :metric
-  [{:keys [id]}]
-  (-> (db/select-one Metric :id id) mi/can-read?))
+  [o]
+  (mi/can-read? o))
```
Comment (camsaul) [inline src/metabase/api/search.clj:369]: We're using different names all over the place to mean "row" or "instance". I've also seen: `entity`, `object`, `obj`, `o`, and the list goes on.

Let's standardize on calling an instance of a Toucan model an _instance_

```suggestion
  [instance]
  (mi/can-read? instance))
```

File: src/metabase/api/search.clj [hunk 6]
```diff
@@ -364,16 +365,16 @@
...
-  [{:keys [id]}]
-  (-> (db/select-one Segment :id id) mi/can-read?))
+  [o]
+  (mi/can-read? o))
 
 (defmethod check-permissions-for-model :database
-  [{:keys [id]}]
-  (-> (db/select-one Database :id id) mi/can-read?))
+  [o]
+  (mi/can-read? o))
```
Comment (camsaul) [inline src/metabase/api/search.clj:377]:
```suggestion
  [instance]
  (mi/can-read? instance))
```

File: src/metabase/api/search.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/api/search.clj:413]: That is correct. I think this looks fine to me

File: src/metabase/search/config.clj [hunk 7]
```diff
@@ -39,14 +39,14 @@
...
-   "table"      Table
-   "database"   Database})
+  {"dashboard"  :metabase.models.dashboard/Dashboard
+   "metric"     :metabase.models.metric/Metric
+   "segment"    :metabase.models.segment/Segment
+   "card"       :metabase.models.card/Card
+   "dataset"    :metabase.models.card/Card
+   "collection" :metabase.models.collection/Collection
+   "table"      :metabase.models.table/Table
+   "database"   :metabase.models.database/Database})
```
Comment (camsaul) [inline src/metabase/search/config.clj:49]: You definitely don't need to do this

File: test/metabase/api/search_test.clj [hunk 8]
```diff
@@ -609,3 +611,34 @@
...
+                           :metabase.models.metric/Metric        _             {:table_id table-id
+                                                                                :name     "metric count test 1"}
+                           :metabase.models.metric/Metric        _             {:table_id table-id
+                                                                                :name     "metric count test 2"}
+                           :metabase.models.segment/Segment      _             {:table_id table-id
+                                                                                :name     "segment count test 1"}
+                           :metabase.models.segment/Segment      _             {:table_id table-id
+                                                                                :name     "segment count test 2"}
+                           :metabase.models.segment/Segment      _             {:table_id table-id
+                                                                                :name     "segment count test 3"}]
```
Comment (camsaul) [inline test/metabase/api/search_test.clj:638]: You should really just use the model name directly like `Dashboard` and not the full keyword. The full keyword is way too long to type. Maybe in the future we'll choose different shorter keywords like `:model/Dashboard` but for now just use `Dashboard`

File: test/metabase/api/search_test.clj (hunk 1 above)
Comment (camsaul) [reply test/metabase/api/search_test.clj:644]: 💪

File: test/metabase/api/search_test.clj (hunk 8 above)
Comment (qnkhuat) [reply test/metabase/api/search_test.clj]: gotcha, I thought we were going to deprecate that stuff.

**Pull request merged**